app/routes/crate/version.js

```diff
@@ -25,14 +28,37 @@ export default class VersionRoute extends Route {
     let version;
     let requestedVersion = params.version_num;
     if (requestedVersion) {
-      version = versions.find(version => version.num === requestedVersion);
-      if (!version) {
-        let title = `${crate.name}: Version ${requestedVersion} not found`;
-        return this.router.replaceWith('catch-all', { transition, title });
+      try {
+        version =
+          crate.loadedVersionsByNum.get(requestedVersion) ??
+          (await this.store.queryRecord('version', {
+            name: crate.id,
+            num: requestedVersion,
+          }));
+      } catch (error) {
+        if (error instanceof NotFoundError) {
+          let title = `${crate.name}: Version ${requestedVersion} not found`;
+          return this.router.replaceWith('catch-all', { transition, title });
+        } else {
+          let title = `${crate.name}: Failed to load version data`;
+          return this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
+        }
       }
     } else {
       let { default_version } = crate;
-      version = versions.find(version => version.num === default_version);
+      version =
+        crate.loadedVersionsByNum.get(default_version) ??
+        (await this.store
+          .queryRecord('version', {
+            name: crate.id,
+            num: default_version,
+          })
+          .catch(error => {
+            if (!(error instanceof NotFoundError)) {
+              let title = `${crate.name}: Failed to load version data`;
+              this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
+            }
+          }));
 
       if (!version) {
```

Comment:
Since we already have `default_version`, do we still want this fallback here?

Reply:
More specifically, I'm talking about this part https://github.com/rust-lang/crates.io/blob/859959f065fa873da96c494fcbed44fc70fcb727/app/routes/crate/version.js#L63-L68

Reply:
no, I think we can get rid of that and replace it with an error screen